src/workerd/api/streams/readable-source.c++

```diff
@@ -0,0 +1,817 @@
+  // Used for tracking if this body was ever used.
+  bool wasRead = false;
+};
+
+// A ReadableStreamSource implementation that lazily wraps an innner Gzip or Brotli
+// encoded AsyncInputStream when the first read() is called, or when pumpTo is called,
+// the encoding will be selectively and lazily applied to the inner stream.
+class EncodedAsyncInputStream final: public ReadableStreamSourceImpl {
+ public:
+  EncodedAsyncInputStream(kj::Own<kj::AsyncInputStream> inner, rpc::StreamEncoding encoding)
+      : ReadableStreamSourceImpl(kj::mv(inner), encoding) {}
+
+  // Read bytes in identity encoding. If the stream is not already in identity encoding, it will be
+  // converted to identity encoding via an appropriate stream wrapper.
+  kj::Promise<size_t> read(kj::ArrayPtr<kj::byte> buffer, size_t minBytes) override {
+    try {
+      co_return co_await ReadableStreamSourceImpl::read(buffer, minBytes);
+    } catch (...) {
+      auto exception = kj::getCaughtExceptionAsKj();
+      KJ_IF_SOME(translated,
+          translateKjException(exception,
+              {
+                {"gzip compressed stream ended prematurely"_kj,
+                  "Gzip compressed stream ended prematurely."_kj},
+                {"gzip decompression failed"_kj, "Gzip decompression failed."},
+                {"brotli state allocation failed"_kj, "Brotli state allocation failed."},
+                {"invalid brotli window size"_kj, "Invalid brotli window size."},
+                {"invalid brotli compression level"_kj, "Invalid brotli compression level."},
+                {"brotli window size too big"_kj, "Brotli window size too big."},
+                {"brotli decompression failed"_kj, "Brotli decompression failed."},
+                {"brotli compression failed"_kj, "Brotli compression failed."},
+                {"brotli compressed stream ended prematurely"_kj,
+                  "Brotli compressed stream ended prematurely."},
+              })) {
+        kj::throwFatalException(kj::mv(translated));
+      } else {
+        kj::throwFatalException(kj::mv(exception));
+      }
+    }
+  }
+
+  kj::Maybe<Tee> tryTee(kj::Maybe<size_t> maybeLimit) override {
+    // Note that if we haven't called read() yet, then the inner stream is still
+    // in its original encoding. If read() has been called, however, then the inner
+    // stream will be wrapped and will be in identity encoding.
+    static constexpr size_t kMax = kj::maxValue;
+    auto& inner = KJ_ASSERT_NONNULL(getState().tryGet<kj::Own<kj::AsyncInputStream>>());
+    auto tee = kj::newTee(kj::mv(inner), maybeLimit.orDefault(kMax));
+    return Tee{
+      .branch1 =
+          kj::heap<EncodedAsyncInputStream>(wrapTeeBranch(kj::mv(tee.branches[0])), getEncoding()),
+      .branch2 =
+          kj::heap<EncodedAsyncInputStream>(wrapTeeBranch(kj::mv(tee.branches[1])), getEncoding()),
+    };
+  }
+
+  kj::AsyncInputStream& ensureIdentityEncoding(kj::Own<kj::AsyncInputStream>& inner) override {
```

Comment:
`ensureIdentityEncoding()` seems kind of tricky to use. It may or may not move its argument.

This is kind of a hot take, but IMO the correct way to express "may or may not move its argument" is with an rvalue reference parameter (as opposed to "definitely moves its argument", which would be just a value parameter).

If `ensureIdentityEncoding()` took an rvalue reference, that would force a `kj::mv(inner)` everywhere it's called, which would make it clear that we cannot count on `inner` being valid after `ensureIdentiyEncoding()` is called. The call sites appear to be carefully written to account for that anyway; this would just make it more obvious.

Failing that, more comments explaining what's going on, both here and at this function's call sites, would be helpful.